src/tree.rs

```diff
@@ -202,66 +181,78 @@ impl MerkleCrhOrchardOutput {
+impl std::cmp::PartialEq for MerkleHashOrchard {
     fn eq(&self, other: &Self) -> bool {
         self.0.ct_eq(&other.0).into()
     }
 }
 
 /// This instance should only be used for hash table key comparisons.
-impl std::cmp::Eq for MerkleCrhOrchardOutput {}
+impl std::cmp::Eq for MerkleHashOrchard {}
 
 /// This instance should only be used for hash table key hashing.
-impl std::hash::Hash for MerkleCrhOrchardOutput {
+impl std::hash::Hash for MerkleHashOrchard {
     fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
         <Option<pallas::Base>>::from(self.0)
             .map(|b| b.to_bytes())
             .hash(state)
     }
 }
 
-impl ConditionallySelectable for MerkleCrhOrchardOutput {
+impl ConditionallySelectable for MerkleHashOrchard {
     fn conditional_select(a: &Self, b: &Self, choice: Choice) -> Self {
-        MerkleCrhOrchardOutput(pallas::Base::conditional_select(&a.0, &b.0, choice))
+        MerkleHashOrchard(pallas::Base::conditional_select(&a.0, &b.0, choice))
     }
 }
 
-impl Hashable for MerkleCrhOrchardOutput {
+impl Hashable for MerkleHashOrchard {
     fn empty_leaf() -> Self {
-        MerkleCrhOrchardOutput(*UNCOMMITTED_ORCHARD)
+        MerkleHashOrchard(*UNCOMMITTED_ORCHARD)
     }
 
     /// Implements `MerkleCRH^Orchard` as defined in
     /// <https://zips.z.cash/protocol/protocol.pdf#orchardmerklecrh>
+    ///
+    /// The layer with 2^n nodes is called "layer n":
+    ///      - leaves are at layer MERKLE_DEPTH_ORCHARD = 32;
+    ///      - the root is at layer 0.
+    /// `l` is MERKLE_DEPTH_ORCHARD - layer - 1.
+    ///      - when hashing two leaves, we produce a node on the layer above the leaves, i.e.
+    ///        layer = 31, l = 0
+    ///      - when hashing to the final root, we produce the anchor with layer = 0, l = 31.
     fn combine(altitude: Altitude, left: &Self, right: &Self) -> Self {
-        hash_with_l(
-            altitude.into(),
-            Pair {
-                left: left.0,
-                right: right.0,
-            },
+        // MerkleCRH Sinsemilla hash domain.
+        let domain = HashDomain::new(MERKLE_CRH_PERSONALIZATION);
+
+        MerkleHashOrchard(
+            domain
+                .hash(
+                    iter::empty()
+                        .chain(i2lebsp_k(altitude.into()).iter().copied())
+                        .chain(left.0.to_le_bits().iter().by_val().take(L_ORCHARD_MERKLE))
+                        .chain(right.0.to_le_bits().iter().by_val().take(L_ORCHARD_MERKLE)),
+                )
+                .unwrap_or(pallas::Base::zero()),
         )
-        .map(MerkleCrhOrchardOutput)
-        .unwrap_or_else(|| MerkleCrhOrchardOutput(pallas::Base::zero()))
     }
 
     fn empty_root(altitude: Altitude) -> Self {
-        MerkleCrhOrchardOutput(EMPTY_ROOTS[<usize>::from(altitude)])
+        EMPTY_ROOTS[<usize>::from(altitude)]
     }
 }
 
-impl Serialize for MerkleCrhOrchardOutput {
+impl Serialize for MerkleHashOrchard {
```

Comment:
Todo: remove the `Serialize` and `Deserialize` impls before we cut the first release. Non-blocking.